executor/docs/spec.md

```diff
@@ -0,0 +1,276 @@
+# Cook Executor
+
+The document outlines a specification for Cook's custom executor.
+
+It is a modular design allowing users to add functionality by writing reusable, single-responsibility processes. Declarative configuration would determine when then processes would be executed. An HTTP API exposed by the executor itself would enable communication between the executor and the tasks it starts.
+
+
+## Scheduler HTTP API
+
+The following enhancements to the existing Cook scheduler API will:
+
+1. Allow users to specify multiple commands when creating jobs
+2. Allow users to query the status code or codes of a completed job
+
+### Launch a Job
+
+It should be possible to specify multiple commands when launching a job. Any command failure will fail the whole job. In the future, it may be desirable to have finer-grained control over how failure will be handled (for example, ignoring failures from asynchronous monitoring commands).
+
+Request body schema additions:
+
+| JSON Key | Type | Description |
+|----------|------|-------------|
+| commands | array of command objects | the commands to run, as well as information about how they are to be run |
+
+`command` object schema:
+
+| JSON Key | Type | Description |
+|----------|------|-------------|
+| command  | string | the command to run |
+| async | boolean | a flag determining whether the command is run asynchronously.  defaults to `false` |
+| guard | boolean | a flag determining whether failure of the command will stop execution of the whole job. defaults to `false` |
+
+Example `commands`:
+
+
+    commands: [
```

Comment:
> Why is this better than having "pre", "post", "parallel" and "command" commands so that "pre", "post" and "parallel" are add-ons to the current behavior?

Even with `pre`, `post`, and `parallel`, we'd still need the `guard` option (or equivalent). Having one `commands` structure seemed, and I believe makes retrieving exit statuses for all commands more straightforward. Beyond that, they are mostly equivalent.

> Also, please consider the case when I have a script to set certain environment vars before the main command runs.

I have given this some thought. There are a few ways I can think of: 

1. There could be an additional HTTP endpoint for manipulating env variables
2. There could be a `.env` file on disk, to which scripts would add lines like `NAME=value`.
3. There could be a special type of `wrapper` script that's passed the command. It could set any kind of env variables, but it would also be responsible for running the command.

For now, I'll plan on adding the API endpoint, but using a file would allow you to pull in env variables via a `uri`, which might be desirable.